src/test/java/address/unittests/parser/ParserTest.java

```diff
@@ -0,0 +1,137 @@
+package address.unittests.parser;
+
+import address.model.datatypes.person.ReadOnlyViewablePerson;
+import address.model.datatypes.tag.Tag;
+import address.parser.ParseException;
+import address.parser.Parser;
+import address.parser.expr.Expr;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class ParserTest {
+    Parser parser;
+
+    @Before
+    public void setup() {
+        parser = new Parser();
+    }
+
+    @Test
+    public void parser_multipleQualifiers_correctExprProduced() throws ParseException {
+        String filterString = "name:Mueller tag:friends";
+        Expr expr = parser.parse(filterString);
+
+        ReadOnlyViewablePerson readOnlyViewablePerson = getRovPersonMock(1, "John", "Mueller", "", "", "friends");
+
+        assertTrue(expr.satisfies(readOnlyViewablePerson));
+    }
+
+    @Test
+    public void parser_containsNotExprAndMultipleQualifiers_correctExprProduced() throws ParseException {
+        String filterString = "!name:Mueller tag:friends";
+        Expr expr = parser.parse(filterString);
+
+        ReadOnlyViewablePerson personOne = getRovPersonMock(1, "John", "Mueller", "", "", "friends");
+        ReadOnlyViewablePerson personTwo = getRovPersonMock(2, "John", "Tan", "", "", "friends");
+        ReadOnlyViewablePerson personThree = getRovPersonMock(3, "John", "Lee", "", "", "colleagues");
+
+        assertFalse(expr.satisfies(personOne));
+        assertTrue(expr.satisfies(personTwo));
+        assertFalse(expr.satisfies(personThree));
+    }
+
+    @Test
+    public void parser_multipleNotExprAndMultipleQualifiers_correctExprProduced() throws ParseException {
+        String filterString = "!name:Mueller !tag:friends !!city:Singapore";
+        Expr expr = parser.parse(filterString);
+
+        ReadOnlyViewablePerson personOne = getRovPersonMock(1, "John", "Mueller", "", "", "friends");
+        ReadOnlyViewablePerson personTwo = getRovPersonMock(2, "John", "Tan", "", "Singapore", "friends");
+        ReadOnlyViewablePerson personThree = getRovPersonMock(3, "Mull", "Lee", "", "Malaysia", "colleagues");
+        ReadOnlyViewablePerson personFour = getRovPersonMock(4, "Jack", "Lim", "", "Singapore", "colleagues");
+
+        assertFalse(expr.satisfies(personOne));
+        assertFalse(expr.satisfies(personTwo));
+        assertFalse(expr.satisfies(personThree));
+        assertTrue(expr.satisfies(personFour));
```

Comment:
create a utility method that confirms the result contains only the specified persons?